### PR TITLE
API Major refactor of Hierarchy

### PR DESCRIPTION
**Pull request by tractorcow**
Dependency of https://github.com/silverstripe/silverstripe-framework/pull/6792

Rebased on top of https://github.com/silverstripe/silverstripe-admin/pull/19 which should be merged first

### CODE REVIEW

File: code/LeftAndMain.php [hunk 1]
```diff
@@ -1099,11 +1110,11 @@ protected function getSearchFilter()
         // Validate classname
         $filterClass = $params['FilterClass'];
         $filterInfo = new ReflectionClass($filterClass);
-        if (!$filterInfo->implementsInterface('SilverStripe\\Admin\\LeftAndMain_SearchFilter')) {
+        if (!$filterInfo->implementsInterface(LeftAndMain_SearchFilter::class)) {
```
Comment (chillu) [inline code/LeftAndMain.php]: What does this have to do with Hierarchy refactor?

Comment (tractorcow) [inline code/LeftAndMain.php]: I optimised the code to be more performant when getSearchFilter() is requested more often, as the new rewrite does. At the same time I cleaned up code quality.
